src/coreclr/src/System.Private.CoreLib/src/System/Threading/Thread.CoreCLR.cs

```diff
@@ -484,54 +484,22 @@ private static int CalculateOptimalMaxSpinWaitsPerSpinIteration()
 
-        // The upper bits of t_currentProcessorIdCache are the currentProcessorId. The lower bits of
-        // the t_currentProcessorIdCache are counting down to get it periodically refreshed.
-        // TODO: Consider flushing the currentProcessorIdCache on Wait operations or similar
-        // actions that are likely to result in changing the executing core
-        [ThreadStatic]
-        private static int t_currentProcessorIdCache;
-
-        private const int ProcessorIdCacheShift = 16;
-        private const int ProcessorIdCacheCountDownMask = (1 << ProcessorIdCacheShift) - 1;
-        private const int ProcessorIdRefreshRate = 5000;
-
-        private static int RefreshCurrentProcessorId()
-        {
-            int currentProcessorId = GetCurrentProcessorNumber();
-
-            // On Unix, GetCurrentProcessorNumber() is implemented in terms of sched_getcpu, which
-            // doesn't exist on all platforms.  On those it doesn't exist on, GetCurrentProcessorNumber()
-            // returns -1.  As a fallback in that case and to spread the threads across the buckets
-            // by default, we use the current managed thread ID as a proxy.
-            if (currentProcessorId < 0) currentProcessorId = Environment.CurrentManagedThreadId;
-
-            // Add offset to make it clear that it is not guaranteed to be 0-based processor number
-            currentProcessorId += 100;
-
-            Debug.Assert(ProcessorIdRefreshRate <= ProcessorIdCacheCountDownMask);
-
-            // Mask with int.MaxValue to ensure the execution Id is not negative
-            t_currentProcessorIdCache = ((currentProcessorId << ProcessorIdCacheShift) & int.MaxValue) | ProcessorIdRefreshRate;
-
-            return currentProcessorId;
-        }
-
-        // Cached processor id used as a hint for which per-core stack to access. It is periodically
-        // refreshed to trail the actual thread core affinity.
+        // Cached processor id could be used as a hint for which per-core stripe of data to access to avoid sharing.
+        // It is periodically refreshed to trail the actual thread core affinity.
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static int GetCurrentProcessorId()
         {
-            int currentProcessorIdCache = t_currentProcessorIdCache--;
-            if ((currentProcessorIdCache & ProcessorIdCacheCountDownMask) == 0)
-            {
-                return RefreshCurrentProcessorId();
-            }
+            if (s_isProcessorNumberReallyFast)
+                return GetCurrentProcessorNumber();
 
-            return currentProcessorIdCache >> ProcessorIdCacheShift;
+            return ProcessorIdCache.GetCurrentProcessorId();
         }
 
+        // do a fast check and record in a readonly static so that it could become a JIT constant
+        private static readonly bool s_isProcessorNumberReallyFast = ProcessorIdCache.ProcessorNumberSpeedCheck();
```

Comment:
It would be nice to have this field declaration at the top of the file so this initialization would not be "hidden".

https://github.com/dotnet/runtime/blob/master/docs/coding-guidelines/coding-style.md#c-coding-style

> 14. Fields should be specified at the top within type declarations.

Reply:
`System.Threading.Thread` is a dumping ground for many disjoint pieces of functionality. The existing style used in the file is to have the data defined declared locally. Unclear whether moving all data to the top would make it more readable. If you believe that it would make it more readable, we should do it for all to maintain consistent style.

Reply:
I generally have stronger opinion about instance fields. Hard to tell the size of the instance when fields are scattered.
With statics it is often hard to tell whether at the top or near the use is better. 
I will keep the existing style of this file.

---
In reply to: [357740616](https://github.com/dotnet/runtime/pull/467#discussion_r357740616) [](ancestors = 357740616)